tools/serverpod_cli/lib/src/analyzer/entities/entity_dependency_resolver.dart

```diff
@@ -291,19 +291,37 @@ class EntityDependencyResolver {
     } else {
       var foreignFields = referenceClass.fields.where((field) {
         var fieldRelation = field.relation;
-        if (fieldRelation is! ForeignRelationDefinition) return false;
-        return fieldRelation.parentTable == classDefinition.tableName &&
-            fieldRelation.name == relation.name;
+        if (!(fieldRelation is UnresolvedObjectRelationDefinition ||
+            fieldRelation is ForeignRelationDefinition)) return false;
+        return fieldRelation?.name == relation.name;
       });
 
-      if (foreignFields.isNotEmpty) {
-        fieldDefinition.relation = ListRelationDefinition(
-          name: relation.name,
-          fieldName: 'id',
-          foreignFieldName: foreignFields.first.name,
-          nullableRelation: foreignFields.first.type.nullable,
-        );
+      if (foreignFields.isEmpty) return;
+
+      var foreignField = foreignFields.first;
+
+      String? foreignFieldName;
+
+      var foreignRelation = foreignField.relation;
+      if (foreignRelation is ForeignRelationDefinition) {
+        foreignFieldName = foreignField.name;
+      } else if (foreignRelation is UnresolvedObjectRelationDefinition) {
+        foreignFieldName = foreignRelation.fieldName ??
+            _createImplicitForeignFieldName(foreignField.name);
```

Comment:
Question: Why do we need to encode this into the name? Doesn't this need to be in sync with how we generate the code in the code generator now?

For the list we set the `implicitForeignField: true` so that the formatting can be handled in one place (the code generator). Would that be possible here as well?

Reply:
refactor name to: `_createImplicitForeignIdFieldName(...);`